src/node.h

```diff
@@ -0,0 +1,58 @@
+#pragma once
+#include<iostream>
+
+template<class ValueType>
+class Node
+{
+private:
+    ValueType value;
+
+public:
+    Node* parent;
+    Node* left;
+    Node* right;
+
+    Node(ValueType v);
+    Node(ValueType v, Node<ValueType>* p);
+    ~Node() = default;
+
+    ValueType getValue();
+    void setValue(ValueType v);
+
+    friend std::ostream& operator<<(std::ostream& str, Node<ValueType>& n) {
+        return (str << n.value);
+    };
+    friend std::ostream& operator<<(std::ostream& str, Node<ValueType>* n) {
+        return (str << n->value);
+    };
+};
+
+// Constructor with value
+template<class ValueType>
+Node<ValueType>::Node(ValueType v) : value(v), parent(nullptr), left(nullptr), right(nullptr) {}
+
+// Constructor with value and pointer to parent node
+template<class ValueType>
+Node<ValueType>::Node(ValueType v, Node<ValueType>* p) : value(v), parent(p), left(nullptr), right(nullptr) {
+    if (parent->left == nullptr) {
+        parent->left = this;
+    }
+    else if (parent->right == nullptr) {
+        parent->right = this;
+    }
+    else {
+        throw "Added 3rd child to a Node";
```

Comment:
Rzucając wyjątek nie wywalimy aplikacji, przez co aplikacja nie przestanie działać?

Reply:
Nie, jak nie zrobisz try i catch to kompletnie nic się nie stanie. Poza tym jeżeli dojdzie do dodania 3 childa to tracimy wskaźnik do jednego, więc jest wyciek pamięci.

Reply:
No właśnie, a nie lepiej może zrobić konstruktor bez tego parametru p, żeby nie dopuścić do sytuacji gdy może zaistnieć próba dodania 3 childa?

Reply:
Widzę, że parent i obydwa childy są publiczne. To może zarządzanie Nodem zostawmy już Heapowi. Tzn. Heap gdzie trzeba podłączy parenta

Reply:
I tak i tak trzeba uważać żeby nie dodać 3 childa. Jak wolicie tego pilnować przy dodawaniu to mogę usunąć. Tak przynajmniej samo łączenie nodów jest trochę wydzielone i nie będzie na pewno sytuacji że dziecko wie o rodzicu, a rodzic o dziecku już nie.